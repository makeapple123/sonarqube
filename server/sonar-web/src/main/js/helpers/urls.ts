/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { stringify } from 'querystring';
import { isShortLivingBranch } from './branches';
import { getProfilePath } from '../apps/quality-profiles/utils';
import { Branch } from '../app/types';

interface Query {
  [x: string]: string | undefined;
}

interface Location {
  pathname: string;
  query?: Query;
}

export function getBaseUrl(): string {
  return (window as any).baseUrl;
}

/**
 * Generate URL for a component's home page
 */
export function getComponentUrl(componentKey: string, branch?: string): string {
  const branchQuery = branch ? `&branch=${encodeURIComponent(branch)}` : '';
  return getBaseUrl() + '/dashboard?id=' + encodeURIComponent(componentKey) + branchQuery;
}

export function getProjectUrl(key: string, branch?: string): Location {
  return { pathname: '/dashboard', query: { id: key, branch } };
}

export function getProjectBranchUrl(key: string, branch: Branch): Location {
  if (isShortLivingBranch(branch)) {
    return {
      pathname: '/project/issues',
      query: { branch: branch.name, id: key, resolved: 'false' }
    };
  } else if (!branch.isMain) {
    return { pathname: '/dashboard', query: { branch: branch.name, id: key } };
  } else {
    return { pathname: '/dashboard', query: { id: key } };
  }
}

/**
 * Generate URL for a global issues page
 */
export function getIssuesUrl(query: Query): Location {
  return { pathname: '/issues', query };
}

/**
 * Generate URL for a component's issues page
 */
export function getComponentIssuesUrl(componentKey: string, query?: Query): Location {
  return { pathname: '/project/issues', query: { ...query || {}, id: componentKey } };
}

export function getComponentIssuesUrlAsString(componentKey: string, query?: Query): string {
  const path = getComponentIssuesUrl(componentKey, query);
  return `${getBaseUrl()}${path.pathname}?${stringify(path.query)}`;
}

/**
 * Generate URL for a component's drilldown page
 */
export function getComponentDrilldownUrl(
  componentKey: string,
  metric: string,
  branch?: string
): Location {
  return { pathname: '/component_measures', query: { id: componentKey, metric, branch } };
}

/**
 * Generate URL for a component's measure history
 */
export function getComponentMeasureHistory(
  componentKey: string,
  metric: string,
  branch?: string
): Location {
  return {
    pathname: '/project/activity',
    query: { id: componentKey, graph: 'custom', custom_metrics: metric, branch }
  };
}

/**
 * Generate URL for a component's permissions page
 */
export function getComponentPermissionsUrl(componentKey: string): Location {
  return { pathname: '/project_roles', query: { id: componentKey } };
}

/**
 * Generate URL for a quality profile
 */
export function getQualityProfileUrl(
  name: string,
  language: string,
  organization?: string | null
): Location {
  return getProfilePath(name, language, organization);
}

export function getQualityGateUrl(key: string, organization?: string | null): Location {
  return {
    pathname: getQualityGatesUrl(organization).pathname + '/show/' + encodeURIComponent(key)
  };
}

export function getQualityGatesUrl(organization?: string | null): Location {
  return {
    pathname:
      (organization ? '/organizations/' + encodeURIComponent(organization) : '') + '/quality_gates'
  };
}

/**
 * Generate URL for the rules page
 */
export function getRulesUrl(query: { [x: string]: string }, organization?: string | null): string {
  const path = organization ? `/organizations/${organization}/rules` : '/coding_rules';

  if (query) {
    const serializedQuery = Object.keys(query)
      .map(criterion => `${encodeURIComponent(criterion)}=${encodeURIComponent(query[criterion])}`)
      .join('|');

    // return a string (not { pathname }) to help react-router's Link handle this properly
    return path + '#' + serializedQuery;
  }

  return path;
}

/**
 * Generate URL for the rules page filtering only active deprecated rules
 */
export function getDeprecatedActiveRulesUrl(query = {}, organization?: string | null): string {
  const baseQuery = { activation: 'true', statuses: 'DEPRECATED' };
  return getRulesUrl({ ...query, ...baseQuery }, organization);
}

export function getProjectsUrl(): string {
  return getBaseUrl() + '/projects';
}

export function getMarkdownHelpUrl(): string {
  return getBaseUrl() + '/markdown/help';
}
