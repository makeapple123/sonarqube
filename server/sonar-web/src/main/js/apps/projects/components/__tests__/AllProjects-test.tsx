/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
jest.mock('../ProjectsListContainer', () => ({
  default: function ProjectsListContainer() {
    return null;
  }
}));

jest.mock('../ProjectsListFooterContainer', () => ({
  default: function ProjectsListFooterContainer() {
    return null;
  }
}));
jest.mock('../PageHeaderContainer', () => ({
  default: function PageHeaderContainer() {
    return null;
  }
}));

jest.mock('../PageSidebar', () => ({
  default: function PageSidebar() {
    return null;
  }
}));

jest.mock('../../../../helpers/storage', () => ({
  getSort: () => null,
  getView: jest.fn(() => null),
  getVisualization: () => null,
  saveSort: jest.fn(),
  saveView: jest.fn(),
  saveVisualization: jest.fn()
}));

import * as React from 'react';
import { mount, shallow } from 'enzyme';
import AllProjects from '../AllProjects';
import { getView, saveSort, saveView, saveVisualization } from '../../../../helpers/storage';

beforeEach(() => {
  (getView as jest.Mock<any>).mockImplementation(() => null);
  (saveSort as jest.Mock<any>).mockClear();
  (saveView as jest.Mock<any>).mockClear();
  (saveVisualization as jest.Mock<any>).mockClear();
});

it('renders', () => {
  const wrapper = shallow(
    <AllProjects
      fetchProjects={jest.fn()}
      isFavorite={false}
      location={{ pathname: '/projects', query: {} }}
    />,
    { context: { router: {} } }
  );
  expect(wrapper).toMatchSnapshot();
  wrapper.setState({ query: { view: 'visualizations' } });
  expect(wrapper).toMatchSnapshot();
});

it('fetches projects', () => {
  const fetchProjects = jest.fn();
  mountRender({ fetchProjects });
  expect(fetchProjects).lastCalledWith(
    {
      coverage: null,
      duplications: null,
      gate: null,
      languages: null,
      maintainability: null,
      new_coverage: null,
      new_duplications: null,
      new_lines: null,
      new_maintainability: null,
      new_reliability: null,
      new_security: null,
      reliability: null,
      search: null,
      security: null,
      size: null,
      sort: null,
      tags: null,
      view: undefined,
      visualization: null
    },
    false,
    undefined
  );
});

it('redirects to the saved search', () => {
  (getView as jest.Mock<any>).mockImplementation(() => 'leak');
  const replace = jest.fn();
  mountRender({}, jest.fn(), replace);
  expect(replace).lastCalledWith({ pathname: '/projects', query: { view: 'leak' } });
});

it('changes sort', () => {
  const push = jest.fn();
  const wrapper = mountRender({}, push);
  wrapper.find('PageHeaderContainer').prop<Function>('onSortChange')('size', false);
  expect(push).lastCalledWith({ pathname: '/projects', query: { sort: 'size' } });
  expect(saveSort).lastCalledWith('size');
});

it('changes perspective to leak', () => {
  const push = jest.fn();
  const wrapper = mountRender({}, push);
  wrapper.find('PageHeaderContainer').prop<Function>('onPerspectiveChange')({ view: 'leak' });
  expect(push).lastCalledWith({
    pathname: '/projects',
    query: { view: 'leak', visualization: undefined }
  });
  expect(saveSort).lastCalledWith(undefined);
  expect(saveView).lastCalledWith('leak');
  expect(saveVisualization).lastCalledWith(undefined);
});

it('updates sorting when changing perspective from leak', () => {
  const push = jest.fn();
  const wrapper = mountRender(
    { location: { pathname: '/projects', query: { sort: 'new_coverage', view: 'leak' } } },
    push
  );
  wrapper.find('PageHeaderContainer').prop<Function>('onPerspectiveChange')({
    view: undefined
  });
  expect(push).lastCalledWith({
    pathname: '/projects',
    query: { sort: 'coverage', view: undefined, visualization: undefined }
  });
  expect(saveSort).lastCalledWith('coverage');
  expect(saveView).lastCalledWith(undefined);
  expect(saveVisualization).lastCalledWith(undefined);
});

it('changes perspective to risk visualization', () => {
  const push = jest.fn();
  const wrapper = mountRender({}, push);
  wrapper.find('PageHeaderContainer').prop<Function>('onPerspectiveChange')({
    view: 'visualizations',
    visualization: 'risk'
  });
  expect(push).lastCalledWith({
    pathname: '/projects',
    query: { view: 'visualizations', visualization: 'risk' }
  });
  expect(saveSort).lastCalledWith(undefined);
  expect(saveView).lastCalledWith('visualizations');
  expect(saveVisualization).lastCalledWith('risk');
});

function mountRender(props: any = {}, push: Function = jest.fn(), replace: Function = jest.fn()) {
  return mount(
    <AllProjects
      fetchProjects={jest.fn()}
      isFavorite={false}
      location={{ pathname: '/projects', query: {} }}
      {...props}
    />,
    { context: { router: { push, replace } } }
  );
}
