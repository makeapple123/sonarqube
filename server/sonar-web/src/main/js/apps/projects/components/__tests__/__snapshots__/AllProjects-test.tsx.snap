// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="layout-page projects-page"
>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <div
    className="layout-page-side-outer"
  >
    <div
      className="layout-page-side projects-page-side"
      style={
        Object {
          "top": 30,
        }
      }
    >
      <div
        className="layout-page-side-inner"
      >
        <div
          className="layout-page-filters"
        >
          <PageSidebar
            isFavorite={false}
            query={Object {}}
            view="overall"
            visualization="risk"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="layout-page-main projects-page-content"
  >
    <div
      className="layout-page-header-panel layout-page-main-header"
    >
      <div
        className="layout-page-header-panel-inner layout-page-main-header-inner"
      >
        <div
          className="layout-page-main-inner"
        >
          <PageHeaderContainer
            isFavorite={false}
            onPerspectiveChange={[Function]}
            onSortChange={[Function]}
            query={Object {}}
            selectedSort="name"
            view="overall"
            visualization="risk"
          />
        </div>
      </div>
    </div>
    <div
      className="layout-page-main-inner"
    >
      <ProjectsListContainer
        cardType="overall"
        isFavorite={false}
        isFiltered={false}
      />
      <ProjectsListFooterContainer
        isFavorite={false}
        query={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`renders 2`] = `
<div
  className="layout-page projects-page"
>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <div
    className="layout-page-side-outer"
  >
    <div
      className="layout-page-side projects-page-side"
      style={
        Object {
          "top": 30,
        }
      }
    >
      <div
        className="layout-page-side-inner"
      >
        <div
          className="layout-page-filters"
        >
          <PageSidebar
            isFavorite={false}
            query={
              Object {
                "view": "visualizations",
              }
            }
            view="visualizations"
            visualization="risk"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="layout-page-main projects-page-content"
  >
    <div
      className="layout-page-header-panel layout-page-main-header"
    >
      <div
        className="layout-page-header-panel-inner layout-page-main-header-inner"
      >
        <div
          className="layout-page-main-inner"
        >
          <PageHeaderContainer
            isFavorite={false}
            onPerspectiveChange={[Function]}
            onSortChange={[Function]}
            query={
              Object {
                "view": "visualizations",
              }
            }
            selectedSort="name"
            view="visualizations"
            visualization="risk"
          />
        </div>
      </div>
    </div>
    <div
      className="layout-page-main-inner"
    >
      <Connect(Visualizations)
        visualization="risk"
      />
    </div>
  </div>
</div>
`;
